--- a/include/net/dsa.h	2024-07-05 13:57:32.295830092 +0900
+++ b/include/net/dsa.h	2024-07-05 13:59:27.469017718 +0900
@@ -656,7 +656,9 @@
 /* Return true if this is the local port used to reach the CPU port */
 static inline bool dsa_is_upstream_port(struct dsa_switch *ds, int port)
 {
-	if (dsa_is_unused_port(ds, port))
+	const struct dsa_port *dp = dsa_to_port(ds, port);
+
+	if (!dp || dp->type == DSA_PORT_TYPE_UNUSED)
 		return false;
 
 	return port == dsa_upstream_port(ds, port);
